src/StepperISR.h

```diff
@@ -90,8 +101,8 @@ struct queue_entry {
 class StepperQueue {
  public:
   struct queue_entry entry[QUEUE_LEN];
-  uint8_t read_idx;  // ISR stops if readptr == next_writeptr
-  uint8_t next_write_idx;
+  volatile uint8_t read_idx;  // ISR stops if readptr == next_writeptr
+  volatile uint8_t next_write_idx;
```

Comment:
This `volatile` is problematic, because it impacts the other architectures. If this is necessary, it should be conditionally compiled for sam only. But I tend to assume, this is not needed.

Why is it important to avoid ?
In the ISR, those two variables are accessed more than once. The `volatile` turns any optimization and for avr ISR service time being as short as possible is crucial.

Why is it not needed ?
For the ISR those two variables are not volatile at all. During ISR execution no one will change it. In the cyclic task/interrupt, the volatile nature is mitigated by 1.disabling the interrupts 2. not reading more than once.

Reply:
Hmmm....My understanding of volatile is quite different, and perhaps its handled differently in the embedded world.  volatile is a hint to the compiler to warn it that it may have been written to by another variable, and to not cache the value.   I didn't test it without, but I did wonder how this ever worked anywhere.  It may even be the gcc for AVR has no concept of variable caching.  All it should ever do though is *hint* the compiler to *always* read the value from the source.  Unfortunately in bigger systems, it is *just a hint* and the compiler does ignore it sometimes.  It should not disable optimizations.  Again, I've spent most of my time in much much much higher powered multiprocessor, kernel threaded systems.  I'd be pretty annoyed though if something so basic as volatile was interpreted differently on smaller architectures.

Reply:
volatile is IMHO not really a hint. It should be a must for the compiler to not cache accordingly declared variables. If a compiler is just ignoring it, then I would be surprised. How can someone then write a reliable device driver ? That’s why most peripheral registers should be defined as volatile. For example a port input register may yield on every read a different value due to changed external signals. Imagine you need to busy wait for a high on an input signal to trigger a time critical action. With a wrongly optimizing compiler this turns into an endless loop and it is impossible to solve this without resorting to assembly code.

In any “parallel” execution environment (just using interrupts), shared variables are critical and just mark those as volatile is just an easy and working way out for the developer, but sacrifices performance a bit. But fave is with one interrupt per step already close to the limits. That’s why I have explained, why I do not see the need in this case and the volatile needs to be removed for at least avr (esp perhaps not) and for sam it’s up to you/your code.

Reply:
The ifndef should be an ifdef, right?

Reply:
I started writing it one way, then implemented it the other....I was doing ifndef and moving the volatile, but reversed it...give me a second...

Reply:
and it turns out, I'm running with that mistake, so I do not have volatile....I had intended to run an all day torture test.  Its up to 12 million pulses without error, and I don't really want to stop it now, so I'll leave my extreme long term test running....guess I'll prove to myself whether it is necessary...